bin/cache_daemon.ml

```diff
@@ -0,0 +1,129 @@
+      | None ->
+        Thread.delay 0.1;
+        loop (x + 1) )
+  in
+  loop 0
+
+let start ~exit_no_client ~foreground ~port_path ~root =
+  let show_endpoint ep = Printf.printf "listening on %s\n%!" ep in
+  let config = { Dune_manager.exit_no_client } in
+  let f started =
+    let started content =
+      if foreground then show_endpoint content;
+      started content
+    in
+    Console.init
+      ( if foreground then
+        Verbose
+      else
+        Quiet );
+    Dune_manager.daemon ~root ~config started
+  in
+  match Daemonize.daemonize ~workdir:root ~foreground port_path f with
+  | Result.Ok Finished -> ()
+  | Result.Ok (Daemonize.Started (endpoint, _)) -> show_endpoint endpoint
+  | Result.Ok (Daemonize.Already_running (endpoint, _)) when not foreground ->
+    show_endpoint endpoint
+  | Result.Ok (Daemonize.Already_running (endpoint, pid)) ->
+    User_error.raise [ Pp.textf "already running on %s (PID %i)" endpoint pid ]
+  | Result.Error reason -> User_error.raise [ Pp.text reason ]
+
+let stop ~port_path =
+  match
+    Result.ok_exn (Dune_manager.check_port_file ~close:false port_path)
+  with
+  | None -> User_error.raise [ Pp.textf "not running" ]
+  | Some (_, pid, fd) ->
+    Unix.kill pid Sys.sigterm;
+    Result.ok_exn
+      (retry
+         ~message:(Printf.sprintf "waiting for daemon to stop (PID %i)" pid)
+         (fun () -> Option.some_if (Fcntl.lock_get fd Fcntl.Write = None) ()))
+
+type mode =
+  | Start
+  | Stop
+
+let modes = [ ("start", Start); ("stop", Stop) ]
+
+let path_conv = ((fun s -> `Ok (Path.of_string s)), Path.pp)
+
+let term =
+  Term.ret
+  @@ let+ mode =
+       Arg.(
+         value
+         & pos 0 (some (enum modes)) None
+         & info [] ~docv:"ACTION"
+             ~doc:
+               (Printf.sprintf "The cache-daemon action to perform (%s)"
+                  (Arg.doc_alts_enum modes)))
+     and+ foreground =
+       Arg.(
+         value & flag
+         & info [ "foreground"; "f" ]
+             ~doc:"Whether to start in the foreground or as a daeon")
+     and+ exit_no_client =
+       let doc = "Whether to exit once all clients have disconnected" in
+       Arg.(
+         value & flag
+         & info [ "exit-no-client" ] ~doc
+             ~env:(Arg.env_var "DUNE_CACHE_EXIT_NO_CLIENT" ~doc))
+     and+ port_path =
+       Arg.(
+         value
+         & opt path_conv (Dune_manager.default_port_file ())
+         & info ~docv:"PATH" [ "port-file" ]
+             ~doc:"The file to read/write the daemon port to/from.")
+     and+ root =
+       Arg.(
+         value
+         & opt path_conv (Dune_memory.default_root ())
+         & info ~docv:"PATH" [ "root" ] ~doc:"Root of the dune cache")
+     in
+     try
+       match mode with
+       | Some Start -> `Ok (start ~exit_no_client ~foreground ~port_path ~root)
+       | Some Stop -> `Ok (stop ~port_path)
+       | None -> `Help (`Pager, Some name)
+     with Failure msg -> `Error (false, msg)
```

Comment:
We shouldn't catch this error. There is already a catch-all error handler in `main.ml`

Reply:
Reading the code I do agree, but computer says no: ```Error: exception Failure("too many retries (100): waiting for daemon to stop (PID 10860)")
Backtrace:
Raised at file "src/stdune/result.ml", line 17, characters 15-22
Called from file "bin/cache_daemon.ml", line 124, characters 24-41
Called from file "vendor/cmdliner/src/cmdliner_term.ml", line 25, characters 19-24
Called from file "vendor/cmdliner/src/cmdliner.ml", line 30, characters 14-21
Called from file "vendor/cmdliner/src/cmdliner.ml", line 146, characters 9-16
Called from file "vendor/cmdliner/src/cmdliner.ml", line 176, characters 18-36
Called from file "vendor/cmdliner/src/cmdliner.ml", line 312, characters 20-46
Called from file "bin/main.ml", line 198, characters 10-51

I must not segfault.  Uncertainty is the mind-killer.  [...]```

Reply:
Ah wait, that's the catchall. But I don't want a frightening assertion message, I just want to print out "couldn't stop the daemon after X tries".

Reply:
What about using `User_error.raise` in `retry` rather than returning `Error (Failure ...)`?

Reply:
Absolutely ; done.